File: examples/src/WPF/DialingVpnEntry/CS/DialingVpnEntry/DispatcherSynchronizingObject.cs
using System;
using System.ComponentModel;
using System.Windows.Threading;

namespace DialingVpnEntryWithWPF
{
    internal class DispatcherSynchronizingObject : ISynchronizeInvoke
    {
        private Dispatcher dispatcher;

        public DispatcherSynchronizingObject(Dispatcher dispatcher)
        {
            this.dispatcher = dispatcher;
        }

        public bool InvokeRequired
        {
            get { return true; }
        }

        public IAsyncResult BeginInvoke(Delegate method, object[] args)
        {
            return new DispatcherAsyncResult(this.dispatcher.BeginInvoke(method, args));
        }

        public object EndInvoke(IAsyncResult result)
        {
            DispatcherAsyncResult asyncResult = result as DispatcherAsyncResult;
            if (asyncResult != null)
            {
                DispatcherOperation dispatcherOperation = asyncResult.AsyncState as DispatcherOperation;
                if (dispatcherOperation != null)
                {
                    return dispatcherOperation.Wait();
                }
            }

            return null;
        }

        public object Invoke(Delegate method, object[] args)
        {
            return this.dispatcher.Invoke(method, args);
        }
    }
}

File: examples/src/WPF/DialingVpnEntry/VB/DialingVpnEntry/DispatcherSynchronizingObject.vb
Imports System.Windows.Threading

Public Class DispatcherSynchronizingObject
    Implements System.ComponentModel.ISynchronizeInvoke

    Private dispatcher As Dispatcher

    Public Sub New(ByVal dispatcher As Dispatcher)
        Me.dispatcher = dispatcher
    End Sub

    Public Function BeginInvoke(ByVal method As System.Delegate, ByVal args() As Object) As System.IAsyncResult Implements System.ComponentModel.ISynchronizeInvoke.BeginInvoke
        Return New DispatcherAsyncResult(Me.dispatcher.BeginInvoke(method, args))
    End Function

    Public Function EndInvoke(ByVal result As System.IAsyncResult) As Object Implements System.ComponentModel.ISynchronizeInvoke.EndInvoke
        Dim asyncResult = CType(result, DispatcherAsyncResult)
        If (asyncResult IsNot Nothing) Then
            Dim dispatcherOperation As DispatcherOperation = CType(asyncResult.AsyncState, DispatcherOperation)
            If (dispatcherOperation IsNot Nothing) Then
                Return dispatcherOperation.Wait()
            End If
        End If

        Return Nothing
    End Function

    Public Function Invoke(ByVal method As System.Delegate, ByVal args() As Object) As Object Implements System.ComponentModel.ISynchronizeInvoke.Invoke
        Return Me.dispatcher.Invoke(method, args)
    End Function

    Public ReadOnly Property InvokeRequired() As Boolean Implements System.ComponentModel.ISynchronizeInvoke.InvokeRequired
        Get
            Return True
        End Get
    End Property
End Class
